drivers/storage/portworx/portworx.go

```diff
@@ -297,92 +304,192 @@ func (p *portworx) updateStorageNodes(
+			msg := fmt.Sprintf("Failed to update StorageNode for nodeID %v: %v", node.Id, err)
+			p.warningEvent(cluster, util.FailedSyncReason, msg)
+			continue
 		}
 
-		err = k8sutil.CreateOrUpdateStorageNode(p.k8sClient, storageNode, ownerRef)
+		err = p.updateStorageNodeStatus(storageNode, node)
 		if err != nil {
-			msg := fmt.Sprintf("Failed to update status for nodeID %v: %v", node.Id, err)
+			msg := fmt.Sprintf("Failed to update StorageNode status for nodeID %v: %v", node.Id, err)
 			p.warningEvent(cluster, util.FailedSyncReason, msg)
 		}
 	}
 
-	nodeStatusList := &corev1alpha1.StorageNodeList{}
-	if err = p.k8sClient.List(context.TODO(), nodeStatusList, &client.ListOptions{}); err != nil {
+	return p.updateRemainingStorageNodes(cluster, currentPxNodes)
+}
+
+func (p *portworx) updateRemainingStorageNodesWithoutError(
+	cluster *corev1alpha1.StorageCluster,
+	currentPxNodes map[string]bool,
+) {
+	if err := p.updateRemainingStorageNodes(cluster, nil); err != nil {
+		logrus.Warn(err)
+	}
+}
+
+func (p *portworx) updateRemainingStorageNodes(
+	cluster *corev1alpha1.StorageCluster,
+	currentPxNodes map[string]bool,
+) error {
+	// Find all k8s nodes where Portworx pods are running
+	pxLabels := p.GetSelectorLabels()
+	portworxPodList := &v1.PodList{}
+	err := p.k8sClient.List(
+		context.TODO(),
+		portworxPodList,
+		&client.ListOptions{
+			Namespace:     cluster.Namespace,
+			LabelSelector: labels.SelectorFromSet(pxLabels),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get list of portworx pods. %v", err)
+	}
+
+	currentPxPodNodes := make(map[string]bool)
+	for _, pod := range portworxPodList.Items {
+		controllerRef := metav1.GetControllerOf(&pod)
+		if controllerRef != nil && controllerRef.UID == cluster.UID && len(pod.Spec.NodeName) != 0 {
+			currentPxPodNodes[pod.Spec.NodeName] = true
+		}
+	}
+
+	storageNodes := &corev1alpha1.StorageNodeList{}
+	if err = p.k8sClient.List(context.TODO(), storageNodes, &client.ListOptions{}); err != nil {
 		return fmt.Errorf("failed to get a list of StorageNode: %v", err)
 	}
 
-	for _, nodeStatus := range nodeStatusList.Items {
-		if _, exists := currentNodes[nodeStatus.Name]; !exists {
+	for _, storageNode := range storageNodes.Items {
+		pxNodeExists := currentPxNodes[storageNode.Name]
+		pxPodExists := currentPxPodNodes[storageNode.Name]
+		if !pxNodeExists && !pxPodExists {
 			logrus.Debugf("Deleting orphan StorageNode %v/%v",
-				nodeStatus.Namespace, nodeStatus.Name)
-			err = p.k8sClient.Delete(context.TODO(), nodeStatus.DeepCopy())
+				storageNode.Namespace, storageNode.Name)
+
+			err = p.k8sClient.Delete(context.TODO(), storageNode.DeepCopy())
 			if err != nil && !errors.IsNotFound(err) {
 				msg := fmt.Sprintf("Failed to delete StorageNode %v/%v: %v",
-					nodeStatus.Namespace, nodeStatus.Name, err)
+					storageNode.Namespace, storageNode.Name, err)
 				p.warningEvent(cluster, util.FailedSyncReason, msg)
 			}
+		} else if !pxNodeExists && pxPodExists {
```

Comment:
Since we are not checking for pxNodeExists == true condition in this function, the update Status() will be called only once for such nodes when updateStorageNodeStatus is called on line 334. Correct ?

Reply:
Yes, that's correct. The updateSpec and updateStatus are only called from there. Just extracted out to make it readable.